public/app/plugins/datasource/prometheus/querycache/QueryCache.ts

```diff
@@ -0,0 +1,222 @@
+    // TODO: align from/to to interval to increase probability of hitting backend cache
+
+    const newFrom = request.range.from.valueOf();
+    const newTo = request.range.to.valueOf();
+
+    // only cache 'now'-relative queries (that can benefit from a backfill cache)
+    const shouldCache = request.rangeRaw?.to?.toString() === 'now';
+
+    // all targets are queried together, so we check for any that causes group cache invalidation & full re-query
+    let doPartialQuery = shouldCache;
+    let prevTo: TimestampMs;
+
+    // pre-compute reqTargSigs
+    const reqTargSigs = new Map<TargetIdent, TargetSig>();
+    request.targets.forEach((targ) => {
+      let targIdent = `${request.dashboardUID}|${request.panelId}|${targ.refId}`;
+      let targExpr = interpolateString(targ.expr);
+      let targSig = `${targExpr}|${request.intervalMs}|${JSON.stringify(request.rangeRaw ?? '')}`; // ${request.maxDataPoints} ?
+
+      reqTargSigs.set(targIdent, targSig);
+    });
+
+    // figure out if new query range or new target props trigger full cache invalidation & re-query
+    for (const [targIdent, targSig] of reqTargSigs) {
+      let cached = this.cache.get(targIdent);
+      let cachedSig = cached?.sig;
+
+      if (cachedSig !== targSig) {
+        doPartialQuery = false;
+      } else {
+        // only do partial queries when new request range follows prior request range (possibly with overlap)
+        // e.g. now-6h with refresh <= 6h
+        prevTo = cached?.prevTo ?? Infinity;
+        doPartialQuery = newTo > prevTo && newFrom <= prevTo;
+      }
+
+      if (!doPartialQuery) {
+        break;
+      }
+    }
+
+    // console.log(`${doPartialQuery ? 'partial' : 'full'} query`);
+
+    if (doPartialQuery) {
+      // 10m re-query overlap
+      const requeryLastMs = 10 * 60 * 1000;
+
+      // clamp to make sure we don't re-query previous 10m when newFrom is ahead of it (e.g. 5min range, 30s refresh)
+      let newFromPartial = Math.max(prevTo! - requeryLastMs, newFrom);
+      request.originalQueryDelta = newFromPartial - newFrom;
+
+      // console.log(`query previous ${(newTo - newFromPartial) / 1000 / 60} mins`);
+
+      // align the query to the interval to increase the chance of hitting the backend cache
+      // ASK: how to test if this actually has desired effect?
+      // ASK: does the backend already do this anyways? do we need to do it here?
+      // newFromPartial = incrRoundDn(newFromPartial, request.intervalMs);
+      // newToPartial = incrRoundDn(newTo, request.intervalMs);
+
+      // modify to partial query
+      request = {
+        ...request,
+        range: {
+          ...request.range,
+          from: dateTime(newFromPartial),
+          to: dateTime(newTo),
+        },
+      };
+    } else {
+      reqTargSigs.forEach((targSig, targIdent) => {
+        this.cache.delete(targIdent);
+
+        // TODO: figure out how to purge caches of targets that have permanently stopped making queries
```

Comment:
This might need to be triggered by the react application, when a dashboard is removed or disabled